### ISSUE TITLE
Suggest API's for middleware registration of HTML serving based on JSON

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3251]
Today, stateful Starcounter web applications normally serve JSON to clients.

If the JSON has a special property "Html", and the code-behind class derive `Page` or `Partial`, Starcounter render the referenced HTML, using it as a template, populating it with the returned JSON.

Instead of this implicit model, we aim to move towards one that is based on middleware explicitly enabled or provided by applications.

The aim is to

* reduce hidden "magic"
* provide a simple and understandable request pipeline aligned with well-known patterns

The goals are to

1. remove Page and Partial classes
2. have applications explicitly instruct where HTML is available when the client ask for HTML for a given JSON resource, using (optionally built-in) middleware
3. have applications explicitly enable the "blending" using middleware

This issue is particularly about [2]. I'll take the responsibility for it to be realized by doing research and coming up with suggestions, and then we'll make it into a work issue after we have something we are satisfied with.

/cc @warpech @Starcounter-Jack @dan31

**Comment by per-samuelsson** [comment]
A few general guidelines for the upcoming middleware API's:

1. Aligned with future DNX / ASP.NET 5 stack
2. Middleware handlers should be chained, with handlers being responsible for invoking the next handler in the chain (or terminating if desired)
3. Normal middleware is app-specific, not host-specific ("app isolation")
4. The pipeline is executed in the order it is registered by the application

**Comment by per-samuelsson** [comment]
More design guidelines:

1. Middleware should ultimately be provided with a context that include at least request, response, session, application info.
2. Built-in middleware should support options / customization; custom middleware should be presented with a straight-forward way to support the same.

**Comment by per-samuelsson** [comment]
Another criteria for 2.0: not to intrusive, but rather something that can be used to *complement* (as in "build upon and abstract away") already existing middleware API's. With that: no switch to async request delegate design (we'll take that step in DNX-version), and possibly no explicit chained invocation (i.e. instead have the host handle invoking middleware in a chain).

**Comment by per-samuelsson** [comment]
I haven't been able to spend as much time on this as I'd like, being busy with mostly replication, but now I think I have figured out both how the basic flow of the current server works, how request and response (middleware) filters are currently implemented, and a rough suggestion of how we could design something with not to much effort that are in line with most stuff I've outlined in previous comments.

I hope I'll have time to summarize this over the weekend and/or can start working on it within days, assuming no one have strong objections to the proposal obviously.

**Comment by per-samuelsson** [comment]
I'm having bit of a trouble assessing exactly how the `Page.Html` (or rather, the JSON.Html) is accessed in the bigger context. 

I've learned now how, when constructing the response, the resource abstraction (`IResource`) is asked to transform itself into the preferred MIME type, and that fits together pretty good. When the `Page` class is activated, the referenced HTML fetched and it's bytes returned.

But I've also started to suspect this is not the only place where this `Html` property is fetched? Is it also fetched by some client code (i.e. code that knows about it by name)?

And what about [this line in StandardJsonSerializer](https://github.com/Starcounter/Starcounter/blob/develop/src/Starcounter.XSON/Serializer/Json/StandardJsonSerializer.cs#L281)? Is that a reference to the `Page` property too?

**Comment by chrhol** [comment]
I was going to answer sooner, but got a nasty cold that left me out of the loop for some days :)

> And what about this line in StandardJsonSerializer? Is that a reference to the Page.Html property too?

There are currently some hacks and extra functionality in the json serializer to support blending and sending juicy layouts to the client inside the viewmodel. These were added (as a lot of weird stuff usually are) when there was no time to do anything else and we really needed to get it working. There been some discussions in #2902 but we haven't spent a lot of time on it yet.

The serializer namespaces each partial and adds an extra `Html`-property (and some other stuff) outside nodes that are merged, that includes the html filepaths for all partials.

You can see in [this example](https://github.com/StarcounterPrefabs/Launcher/wiki/Merged-JSON-Response) how it looks. Currently the json in the section **Required for basic layout management** is returned. Only the outer properties are added by the serializer, `Html`, `AppName`, `PartialId` and `layout`. The html filepaths are read using the `IResource` interface.

If the `IResource` interface is still going to be used I think the changes in this issue should not affect this code, but we want to get rid of all hacks and it's not working in all cases so maybe it's a good time to fix this as well now when were changing how partials/pages works.

**Comment by per-samuelsson** [comment]
Wow, it's clearly quite a lot I don't have any clue about in this request processing pipeline, and the more I look at it, the more I get the feeling I'm not really skilled enough to pull this off (i.e. switching to make HTML-provisioning being middleware-based, rather than something built-in by default).

I really appreciate you took the time to write that answer, @chrhol, since I want to learn how it all fit together. Thanks.

I'm just finishing of introducing those new unified middleware API's, so I still haven't done anything more concrete on trying to achieve the above. Just a lot of code-reviewing, and testing. :smile: I'll see what I can make of all that new info you provided, and then we can catch up to see if I should try somehow to pull it off, or we should wait a little and make a little redesign first.

**Comment by per-samuelsson** [comment]
More things I'm considering:

1. Keep the MVVM viewmodel + view provision smooth and simple, but at the same time
2. Attract the REST purists

Prios in that order, but trying to balance for both.

**Comment by per-samuelsson** [comment]
Finally a first draft proposal done. I wrapped it up in an article that I plan to contribute as a blog post, once we can get to a consensus on the details.

The whole article is temporarily hosted here:
https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers

**TL;DR**
I suggest two complementary approaches for returning multiple representations for a single resource in general, with some JSON view model + HTML view sugaring. 

The **recommended** pattern is based on the new middleware API's and a built-in middleware MIME provider (see [article](https://github.com/per-samuelsson/Misc/wiki/Demystifying-request-handlers) for details):

`Author.json`
```json
{
  "Html": "/author.html",
  "Name": "Per",
  "Surname": "Samuelsson"
}
```

`Program.cs`
```cs
void Main() {
  var app = Application.Current;

  app.Use(new HtmlFromJsonProvider());

  Handle.GET("/author", () => {
    return new Author();
  });
}
```

The **alternative** pattern is based on a new `Resource` class, that basically provide a mean to respond with *multiple representations* from a *single handler*.

```cs
Handle.GET("/author", () => {
  var json = new Author();
  var html = "<b>Per</b><br/><div>Samuelsson</div>";

  var r = new Resource();
  
  return r.Json(json).Html(() => return UTF8.GetBytes(html));
});
```

Both patterns, as listed here, include sugaring constructs, built on more general underlying ditos, in the attempt to balance simplicty with a non-verbose API, but still be general enough to cover most requirements. Again, turn to the article if you are interested in the details.

**Comment by warpech** [comment]
Very good read! 

Do you propose to introduce two patterns at once? I think we should only implement and support the recommended one.

Any luck making this work with Puppet?

**Comment by per-samuelsson** [comment]
> Very good read!

Thanks!

> Do you propose to introduce two patterns at once? I think we should only implement and support the recommended one.

My idea was to introduce both at once, yes. But it might be wise just to start with the recommended one, and see if there is actually a need for the "inline" version.

> Any luck making this work with Puppet?

Haven't had much time to do more than simple mocking. The "spec" itself has been the main focus, plus all other stuff. But I did sit down for a couple of hours with @chrhol and as I understand it, I'll need help to get the details working, once we decide finally on it and take it on.
